src/physics/em/EnergyLossDistribution.i.hh

```diff
@@ -0,0 +1,362 @@
+                      / (w - mat.log_mean_excitation_energy().value());
+                for (int i : range(2))
+                {
+                    // Excitation macroscopic cross section (PRM Eq. 7.10)
+                    xs_exc[i] = c * params.oscillator_strength[i]
+                                * (w - params.log_binding_energy[i])
+                                / params.binding_energy[i];
+                }
+            }
+            else
+            {
+                xs_exc[0] = mean_loss * (1 - EnergyLossDistribution::rate())
+                            / params.binding_energy[0];
+            }
+
+            // Scale the binding energy and macroscopic cross section (i.e.,
+            // the mean number of excitations)
+            real_type scaling = 4;
+            if (xs_exc[0] < EnergyLossDistribution::exc_thresh())
+            {
+                scaling = real_type(0.5)
+                          + (scaling - real_type(0.5))
+                                * std::sqrt(
+                                    xs_exc[0]
+                                    / EnergyLossDistribution::exc_thresh());
+            }
+            binding_energy[0] *= scaling;
+            xs_exc[0] /= scaling;
+        }
+    }
+
+    // Calculate the ionization macroscopic cross section (PRM Eq. 7.11)
+    constexpr real_type e_0
+        = EnergyLossDistribution::ionization_energy().value();
+    real_type xs_ion = mean_loss * (max_energy_ - e_0)
+                       / (max_energy_ * e_0 * std::log(max_energy_ / e_0));
+    if (xs_exc[0] + xs_exc[1] > 0)
+    {
+        // The contribution from excitation is nonzero, so scale the ionization
+        // cross section accordingly
+        xs_ion *= EnergyLossDistribution::rate();
+    }
+
+    // Calculate actual energy loss from the loss contributions from excitation
+    // and ionization
+    const real_type result
+        = loss_scaling
+          * (this->sample_excitation_loss(xs_exc, binding_energy, rng)
+             + this->sample_ionization_loss(xs_ion, rng));
+
+    CELER_ENSURE(result > 0);
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the energy loss contribution from excitation for the Urban model.
+ */
+template<class Engine>
+CELER_FUNCTION real_type EnergyLossDistribution::sample_excitation_loss(
+    Real2 xs, Real2 binding_energy, Engine& rng) const
+{
+    real_type result = 0;
+
+    // Mean and variance for fast sampling from Gaussian
+    real_type mean     = 0;
+    real_type variance = 0;
+
+    for (int i : range(2))
+    {
+        if (xs[i] > EnergyLossDistribution::max_collisions())
+        {
+            // WHen the number of collisions is large, use faster approach
+            // of sampling from a Gaussian
+            mean += xs[i] * binding_energy[i];
+            variance += xs[i] * ipow<2>(binding_energy[i]);
+        }
+        else if (xs[i] > 0)
```

Comment:
It looks like the corresponding G4 code just has a plain `else`. What happens if `xs[i] <= 0`?

Reply:
This is because our Poisson distribution expects lambda > 0 while G4Poisson does not. If the excitation cross section is zero, this should just return zero loss.